File: 2015/22/solution.php
<?php

// TODO: INCORRECT RESULTS

function battle(
    int $boss_hp, int $boss_damage, bool $hard_mode = false,
    int $hit_points = 100, int $mana_left = 500,
    int $mana_spent = 0, bool $player_turn = true,
    int $shield = 0, int $poison = 0, int $recharge = 0,
    int &$result = PHP_INT_MAX
): int
{
    if ($hard_mode && $player_turn) {
        $hit_points--;
    }

    if ($hit_points <= 0) {
        return $result;
    }

    if ($shield > 0) {
        $shield--;
    }

    if ($poison > 0) {
        $poison--;
        $boss_hp -= 3;
    }

    if ($recharge > 0) {
        $recharge--;
        $mana_left += 101;
    }

    if ($boss_hp <= 0) {
        return $mana_spent;
    }

    if ($player_turn) {
        // Magic Missile
        $cost = 53;
        if ($mana_left >= $cost && $mana_spent + $cost < $result) {
            $result = min($result, battle(
                $boss_hp - 4, $boss_damage, $hard_mode,
                $hit_points, $mana_left - $cost,
                $mana_spent + $cost, !$player_turn,
                $shield, $poison, $recharge,
                $result
            ));
        }

        // Drain
        $cost = 73;
        if ($mana_left >= $cost && $mana_spent + $cost < $result) {
            $result = min($result, battle(
                $boss_hp - 2, $boss_damage, $hard_mode,
                $hit_points + 2, $mana_left - $cost,
                $mana_spent + $cost, !$player_turn,
                $shield, $poison, $recharge,
                $result
            ));
        }

        // Shield
        $cost = 113;
        if ($mana_left >= $cost && !$shield && $mana_spent + $cost < $result) {
            $result = min($result, battle(
                $boss_hp, $boss_damage, $hard_mode,
                $hit_points, $mana_left - $cost,
                $mana_spent + $cost, !$player_turn,
                6, $poison, $recharge,
                $result
            ));
        }

        // Poison
        $cost = 173;
        if ($mana_left >= $cost && !$poison && $mana_spent + $cost < $result) {
            $result = min($result, battle(
                $boss_hp, $boss_damage, $hard_mode,
                $hit_points, $mana_left - $cost,
                $mana_spent + $cost, !$player_turn,
                $shield, 6, $recharge,
                $result
            ));
        }

        // Recharge
        $cost = 229;
        if ($mana_left >= $cost && !$recharge && $mana_spent + $cost < $result) {
            $result = min($result, battle(
                $boss_hp, $boss_damage, $hard_mode,
                $hit_points, $mana_left - $cost,
                $mana_spent + $cost, !$player_turn,
                $shield, $poison, 5,
                $result
            ));
        }

    } elseif ($hit_points >= 0) {
        // Boss attack
        $result = min($result, battle(
            $boss_hp, $boss_damage, $hard_mode,
            $hit_points - $boss_damage + ($shield ? 7 : 0), $mana_left,
            $mana_spent, !$player_turn,
            $shield, $poison, $recharge,
            $result
        ));
    }

    return $result;
}

$boss_hp     = 0;
$boss_damage = 0;

while (preg_match('/^([\w\s]+): (\d+)$/', fgets(STDIN), $stat)) {
    switch ($stat[1]) {
        case 'Hit Points':
            $boss_hp = intval($stat[2]);
            break;
        case 'Damage':
            $boss_damage = intval($stat[2]);
            break;
    }
}

fwrite(STDOUT, battle($boss_hp, $boss_damage) . PHP_EOL);
fwrite(STDOUT, battle($boss_hp, $boss_damage, true) . PHP_EOL);

File: 2015/22/solution.py
import re
import sys
from enum import Enum
from typing import NamedTuple


class Player(NamedTuple):
    hit_points: int = 50
    mana: int = 500
    mana_used: int = 0


class Boss(NamedTuple):
    hit_points: int
    damage: int


class Effect(Enum):
    MAGIC_MISSILE = 'magic_missile'
    DRAIN = 'drain'
    SHIELD = 'shield'
    POISON = 'poison'
    RECHARGE = 'recharge'


class ActiveEffect(NamedTuple):
    effect: Effect
    turns: int


def read_input() -> Boss:
    hit_points = damage = None

    for line in sys.stdin:
        match = re.match(
            r'(?P<type>Hit Points|Damage): (?P<quantity>\d+)', line.strip()
        )
        if match['type'] == 'Hit Points':
            hit_points = int(match['quantity'])
        if match['type'] == 'Damage':
            damage = int(match['quantity'])

    if hit_points is None or damage is None:
        raise Exception('Invalid input file')

    return Boss(hit_points=hit_points, damage=damage)


def apply_effects(
    player: Player, boss: Boss, effects: list[ActiveEffect]
) -> (Player, Boss, list[ActiveEffect]):
    remaining_effects = []

    for active_effect in effects:
        if active_effect.effect == Effect.POISON:
            boss = boss._replace(hit_points=boss.hit_points - 3)

        if active_effect.effect == Effect.RECHARGE:
            player = player._replace(mana=player.mana + 101)

        if active_effect.turns > 1:
            remaining_effects.append(
                active_effect._replace(turns=active_effect.turns - 1)
            )

    return player, boss, remaining_effects


def player_turn(
    player: Player,
    boss: Boss,
    hard_mode: bool,
    effects: list[ActiveEffect] = [],
    best_so_far: int = sys.maxsize,
):
    if hard_mode:
        player = player._replace(hit_points=player.hit_points - 1)
        if player.hit_points <= 0:
            return None

    # Effects
    (player, boss, effects) = apply_effects(player, boss, effects)

    # Check win condition
    if boss.hit_points <= 0:
        return player.mana_used

    # Check loss condition
    if player.mana < 53:
        return None

    # Player turn
    if player.mana >= 53 and player.mana_used + 53 < best_so_far:
        score = player_attack(
            Effect.MAGIC_MISSILE, player, boss, hard_mode, effects, best_so_far
        )
        if score is not None and score < best_so_far:
            best_so_far = score

    if player.mana >= 73 and player.mana_used + 73 < best_so_far:
        score = player_attack(
            Effect.DRAIN, player, boss, hard_mode, effects, best_so_far
        )
        if score is not None and score < best_so_far:
            best_so_far = score

    if (
        player.mana >= 113
        and player.mana_used + 113 < best_so_far
        and not len(list(filter(lambda x: x.effect == Effect.SHIELD, effects)))
    ):
        score = player_attack(
            Effect.SHIELD, player, boss, hard_mode, effects, best_so_far
        )
        if score is not None and score < best_so_far:
            best_so_far = score

    if (
        player.mana >= 173
        and player.mana_used + 173 < best_so_far
        and not len(list(filter(lambda x: x.effect == Effect.POISON, effects)))
    ):
        score = player_attack(
            Effect.POISON, player, boss, hard_mode, effects, best_so_far
        )
        if score is not None and score < best_so_far:
            best_so_far = score

    if (
        player.mana >= 229
        and player.mana_used + 229 < best_so_far
        and not len(list(filter(lambda x: x.effect == Effect.RECHARGE, effects)))
    ):
        score = player_attack(
            Effect.RECHARGE, player, boss, hard_mode, effects, best_so_far
        )
        if score is not None and score < best_so_far:
            best_so_far = score

    return best_so_far


def player_attack(
    attack: Effect,
    player: Player,
    boss: Boss,
    hard_mode: bool,
    effects: list[ActiveEffect] = [],
    best_so_far: int = sys.maxsize,
):
    if attack == Effect.MAGIC_MISSILE:
        player = player._replace(mana=player.mana - 53, mana_used=player.mana_used + 53)
        boss = boss._replace(hit_points=boss.hit_points - 4)

    elif attack == Effect.DRAIN:
        player = player._replace(
            mana=player.mana - 73,
            mana_used=player.mana_used + 73,
            hit_points=player.hit_points + 2,
        )
        boss = boss._replace(hit_points=boss.hit_points - 2)

    elif attack == Effect.SHIELD:
        player = player._replace(
            mana=player.mana - 113, mana_used=player.mana_used + 113
        )
        effects = effects + [ActiveEffect(effect=Effect.SHIELD, turns=6)]

    elif attack == Effect.POISON:
        player = player._replace(
            mana=player.mana - 173, mana_used=player.mana_used + 173
        )
        effects = effects + [ActiveEffect(effect=Effect.POISON, turns=6)]

    elif attack == Effect.RECHARGE:
        player = player._replace(
            mana=player.mana - 229, mana_used=player.mana_used + 229
        )
        effects = effects + [ActiveEffect(effect=Effect.RECHARGE, turns=5)]

    return boss_turn(player, boss, hard_mode, effects, best_so_far)


def boss_turn(
    player: Player,
    boss: Boss,
    hard_mode: bool,
    effects: list[ActiveEffect] = [],
    best_so_far: int = sys.maxsize,
):
    # Check win condition
    if boss.hit_points <= 0:
        return player.mana_used

    shield = 7 if len(list(filter(lambda x: x.effect == Effect.SHIELD, effects))) else 0

    # Effects
    (player, boss, effects) = apply_effects(player, boss, effects)

    # Check win condition
    if boss.hit_points <= 0:
        return player.mana_used

    # Boss turn
    player = player._replace(
        hit_points=player.hit_points - max(boss.damage - shield, 1)
    )

    # Check loss condition
    if player.hit_points <= 0:
        return None

    return player_turn(player, boss, hard_mode, effects, best_so_far)


boss = read_input()
print(player_turn(player=Player(), boss=boss, hard_mode=False))
print(player_turn(player=Player(), boss=boss, hard_mode=True))
